### PR TITLE
[OPPRO-312] Decouple JVM side JNI codes from backend API and core plan processing code

### PR DESCRIPTION
**Pull request by zhztheplayer**
## What changes were proposed in this pull request?

As mentioned in title. The PR decoupled all JVM side JNI codes from backend API and core plan processing code.

For example, in the past one should implement native "doValidate" code to let Gluten validate a interpreted Substrait plan.
In the patch it becomes optional to developers to choose whether to implement "doValidate" in Java / Scala code, or in C++ native code. 

It's not a main goal of this patch to try to make it possible to write a pure-Java backend. The real goals are:

1. To better adapt to Substrait / Arrow 's cross-platform concept;
2. To make it easier to abstract an independent JNI layer for Gazelle & Velox backend;
3. To make it easier to test / debug Gluten without compiling C++ codes. For example, it should become possible that we can write a "ExampleBackend" without a native library to be loaded. Although futher refactors might be still needed to achieve that;
4. In JVM code structure, to make it easier to distinguish which part is related to native, and which part is not.

## How was this patch tested?

Basically some refactors. Current UTs are enough.

### CODE REVIEW

**Comment by rui-mo** [comment]
This refactor LGTM.

**Review by rui-mo — APPROVED** [review]

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala [hunk 1]
```diff
@@ -87,7 +88,7 @@ class CHIteratorApi extends IIteratorApi with Logging {
     if (index < 3) {
       logDebug(s"The substrait plan for partition $index:\n${substraitPlan.toString}")
     }
-    NativePartition(index, substraitPlan.toByteArray, localFilesNodesWithLocations.head._2)
+    GlutenPartition(index, substraitPlan, localFilesNodesWithLocations.head._2)
```
Comment (zzcclp) [inline backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala:91]: `GlutenPartition` will be serializated to each task, why include `Plan` object in GlutenPartition ? Is it better use `byte[]` and reduce data ?

File: gluten-core/src/main/java/io/glutenproject/row/RowIterator.java [hunk 2]
```diff
@@ -21,11 +21,18 @@
...
 import java.io.IOException;
 
-public class RowIterator {
-  private long nativeHandler = 0L;
+/**
+ * This is a built-in native implementation of {@link BaseRowIterator}. This is a part of the
+ * provided {@link io.glutenproject.vectorized.NativeExpressionEvaluator} implementation as
+ * JNI facility for wiring with native codes.
+ */
+// FIXME can we rename this by adding a "native" prefix to make it clear the iterator is
```
Comment (zzcclp) [inline gluten-core/src/main/java/io/glutenproject/row/RowIterator.java:29]: This class will be removed later, I think it's not needed.

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala (hunk 1 above)
Comment (zhztheplayer) [reply backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala:91]: This is to preserve a structured object rather than bytes for JVM codes. Since technically it might be more reasonable to explicitly serialize the plan only when sending to backend 's native code.

Considering a backend that executes a plan in pure JVM code (or do some extra validations before sending to native side), the current `NativePartition` requires manually deserializing from byte[] to a Java plan object then to execute it. While preserving a Java object in partition may avoid that.

Regarding the data size I'll run TPC-H SF 1K to check if regressions are caused by the patch.

File: gluten-core/src/main/java/io/glutenproject/row/RowIterator.java (hunk 2 above)
Comment (zhztheplayer) [reply gluten-core/src/main/java/io/glutenproject/row/RowIterator.java:29]: 👍

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 3]
```diff
@@ -140,6 +142,8 @@ class GlutenConfig(conf: SQLConf) extends Logging {
 
   // This config is used for deciding whether to load the native library.
   // When false, only Java code will be executed for a quick test.
+  //
+  // FIXME remove / rename this option to emphasize on "debug" rather than "native"
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:146]: 👍

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala (hunk 1 above)
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala:91]: Got it.

**Comment by zhztheplayer** [comment]
No performance regression observed from benchmark result so I am going to merge this.  @zzcclp Please still comment if you found any other issues within the patch. Thanks!
